pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,156 @@
+type UtilizationBasedLimiter struct {
+	services.Service
+
+	logger             log.Logger
+	utilizationScanner utilizationScanner
+
+	// Memory limit in bytes
+	memoryLimit uint64
+	// CPU limit in cores
+	cpuLimit float64
+	// Last CPU time counter
+	lastCPUTime float64
+	// The time of the last update
+	lastUpdate     time.Time
+	movingAvg      *math.EwmaRate
+	limitingReason atomic.String
+}
+
+// NewUtilizationBasedLimiter returns a UtilizationBasedLimiter configured with cpuLimit and memoryLimit.
+func NewUtilizationBasedLimiter(cpuLimit float64, memoryLimit uint64, logger log.Logger) *UtilizationBasedLimiter {
+	// Calculate alpha for a minute long window
+	// https://github.com/VividCortex/ewma#choosing-alpha
+	alpha := 2 / (60/resourceUtilizationUpdateInterval.Seconds() + 1)
+	l := &UtilizationBasedLimiter{
+		logger:      logger,
+		cpuLimit:    cpuLimit,
+		memoryLimit: memoryLimit,
+		// Use a minute long window, each sample being a second apart
+		movingAvg: math.NewEWMARate(alpha, resourceUtilizationUpdateInterval),
+	}
+	l.Service = services.NewTimerService(resourceUtilizationUpdateInterval, l.starting, l.update, nil)
+	return l
+}
+
+// LimitingReason returns the current reason for limiting, if any.
+// If an empty string is returned, limiting is disabled.
+func (l *UtilizationBasedLimiter) LimitingReason() string {
+	return l.limitingReason.Load()
+}
+
+func (l *UtilizationBasedLimiter) starting(_ context.Context) error {
+	p, err := procfs.Self()
+	if err != nil {
+		return errors.Wrap(err, "unable to detect CPU/memory utilization, unsupported platform")
+	}
+
+	l.utilizationScanner = procfsScanner{
+		proc: p,
+	}
+	return nil
+}
+
+func (l *UtilizationBasedLimiter) update(_ context.Context) error {
+	l.compute(time.Now())
+	return nil
+}
+
+func (l *UtilizationBasedLimiter) compute(now time.Time) {
+	cpuTime, memUtil, err := l.utilizationScanner.Scan()
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "failed to get CPU and memory stats", "method",
+			l.utilizationScanner.String(), "err", err.Error())
+		// Disable any limiting, since we can't tell resource utilization
+		l.limitingReason.Store("")
+		return
+	}
+
+	lastUpdate := l.lastUpdate
+	l.lastUpdate = now
+
+	lastCPUTime := l.lastCPUTime
+	l.lastCPUTime = cpuTime
+
+	if lastUpdate.IsZero() {
+		return
+	}
+
+	cpuUtil := (cpuTime - lastCPUTime) / now.Sub(lastUpdate).Seconds()
+	l.movingAvg.Add(int64(cpuUtil * 100))
+	l.movingAvg.Tick()
+	cpuA := float64(l.movingAvg.Rate()) / 100
+
+	var reason string
+	if memUtil >= l.memoryLimit {
+		reason = "memory"
+	} else if cpuA >= l.cpuLimit {
+		reason = "cpu"
+	}
+
+	enable := reason != ""
+	prevEnable := l.limitingReason.Load() != ""
+	if enable == prevEnable {
+		// No change
+		return
+	}
+
+	if enable {
+		level.Info(l.logger).Log("msg", "enabling resource utilization based limiting",
+			"reason", reason, "memory_limit", l.memoryLimit, "cpu_limit", l.cpuLimit)
```

Comment:
[nit] Please add the current CPU and memory utilization too in both logs. I think will be quite useful.

Reply:
Done.